### PR TITLE
Remove Parsing of C++ Exception String

### PR DESCRIPTION
**Pull request by codereport**
This job resolves half of https://github.com/rapidsai/cudf/issues/10632 and a part of https://github.com/rapidsai/cudf/issues/10200.

Currently, in certain places in cuDF Python we parse the string messages returned from C++ [std::exception::what](https://en.cppreference.com/w/cpp/error/exception/what). This should be removed and replaced by returning a more appropriate exception that Cython can automatically handle. If there is no "appropriate" exception, then a custom exception should be created and the Cython machinery should be written to handle it.

### CODE REVIEW

File: cpp/include/cudf/utilities/error.hpp [hunk 1]
```diff
@@ -77,37 +77,59 @@ struct fatal_cuda_error : public cuda_error {
...
  *
+ * Defaults to throwing `cudf::logic_error`, but a custom exception may also be
+ * specified.
+ *
  * Example usage:
  *
  * @code
- * CUDF_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch");
+ * // throws cudf::logic_error
+ * CUDF_EXPECTS(lhs.type() != rhs.type(), "Column type mismatch");
```
Comment (wence-) [inline cpp/include/cudf/utilities/error.hpp:87]: `==` ? (or else this is expecting lhs and rhs to having mismatching types so the error string is backwards?)

File: cpp/include/cudf/utilities/error.hpp [hunk 2]
```diff
@@ -77,37 +77,59 @@ struct fatal_cuda_error : public cuda_error {
...
+ *      "Nested types are not supported.");
  * @endcode
  *
- * @param[in] cond Expression that evaluates to true or false
- * @param[in] reason String literal description of the reason that cond is
- * expected to be true
- * @throw cudf::logic_error if the condition evaluates to false.
+ * @param[in] _condition Expression that evaluates to true or false
+ * @param[in] _expection_type The exception type to throw; must inherit
+ *     `std::exception`. If not specified (i.e. if only two macro
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:96]: Let's add a `static_assert` for this while we're at it.

File: cpp/include/cudf/utilities/error.hpp [hunk 3]
```diff
@@ -77,37 +77,59 @@ struct fatal_cuda_error : public cuda_error {
...
  */
-#define CUDF_FAIL(reason) \
-  throw cudf::logic_error("cuDF failure at: " __FILE__ ":" CUDF_STRINGIFY(__LINE__) ": " reason)
+#define CUDF_FAIL(...)                                       \
+  GET_CUDF_FAIL_MACRO(__VA_ARGS__, CUDF_FAIL_2, CUDF_FAIL_1) \
+  (__VA_ARGS__)
+#define GET_CUDF_FAIL_MACRO(_1, _2, NAME, ...) NAME
+#define CUDF_FAIL_2(_what, _exception_type)      \
+  /*NOLINTNEXTLINE(bugprone-macro-parentheses)*/ \
+  throw _exception_type{"cuDF failure at:" __FILE__ ":" CUDF_STRINGIFY(__LINE__) ": " _what};
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/error.hpp:131]: Maybe for DRY this should just call `CUDF_EXPECTS(false, _exception_type, _what)`?

File: cpp/include/cudf_test/cudf_gtest.hpp [hunk 4]
```diff
@@ -114,8 +114,14 @@ struct TypeList<Types<TYPES...>> {
       exception);                                                   \
   } while (0)
 
-#define CUDF_EXPECT_THROW_MESSAGE(x, msg) \
-  EXPECT_THROW_MESSAGE(x, cudf::logic_error, "cuDF failure at:", msg)
+#define CUDF_EXPECT_THROW_MESSAGE(...)                                     \
```
Comment (jrhemstad) [inline cpp/include/cudf_test/cudf_gtest.hpp:117]: I don't know when these test macros snuck in, but I recall a long time ago advocating strongly against them. I think it's fragile and unimportant for us to be testing for the exact contents of an exceptions error message.

Don't block this PR on removing these, but file it away in your memory banks for future work.

**Comment by vyasr** [comment]
I'm going to close this PR. The changes to our macros have been completed in #12078. I'll move the C++ changes into a new PR along with the associated Python changes needed to address #10632.

**Pull request closed without merging**